### PR TITLE
Delay deciding which cancel scope a Cancelled exception belongs to

### LINKED ISSUE #860: Delay deciding which cancel scope a Cancelled exception belongs to
Right now, whenever we create a `Cancelled` exception, we immediately tag it with the cancel scope that's going to catch it.

There's another option that's briefly mentioned in #606: we could wait until `CancelScope.__exit__` to decide whether a given `Cancelled` exception "belongs" to this scope. The check would be "on the current task's stack, am I the outermost unshielded scope that's in the cancelled state?"

This can behave differently from what we have now in cases where there are multiple cancelled `CancelScope`s on the stack.

I'm not entirely sure if this would actually help with #606, because of complications noted there (hello `KeyboardInterrupt`). But it would probably simplify the cancel scope code in general, in particular `_cancel_no_notify`, and the issues described in this comment: https://github.com/python-trio/trio/pull/835#issuecomment-455107324

### PR DESCRIPTION
**Pull request by oremanj**
Fixes #860.

### CODE REVIEW

File: trio/_core/_exceptions.py [hunk 1]
```diff
@@ -47,10 +47,12 @@ class Cancelled(BaseException):
...
-    :exc:`RuntimeError`. It would not be associated with a cancel scope and thus
-    not be caught by Trio. Use
-    :meth:`cancel_scope.cancel() <trio.CancelScope.cancel>`
-    instead.
+    :exc:`RuntimeError`. Unless raised within the context of a
+    cancelled cancel scope, it would not be caught by Trio and would
+    crash your program. Use :meth:`cancel_scope.cancel()
+    <trio.CancelScope.cancel>` instead, so Trio can maintain the
+    invariant that a :exc:`Cancelled` exception only gets raised when
+    there's a cancelled cancel scope that's intended to catch it.
```
Comment (njsmith) [inline trio/_core/_exceptions.py]: This is probably more details on the internals than we really want people thinking about. How about simplifying down to just: "You cannot raise Cancelled yourself. Attempting to do so will produce a RuntimeError. Use cancel_scope.cancel instead."

**Comment by oremanj** [comment]
The 3.8-dev failure in Travis is unrelated, so I think this is ready.

**Comment by njsmith** [comment]
LGTM. That travis failure is annoying... I just filed a bug on CPython ([bpo-35894](https://bugs.python.org/issue35894)), so hopefully it will be fixed soon. In the mean time though our CI is broken. I guess we need to temporarily disable the 3.8-dev test (or mark it "allowed to fail") until that's fixed?

**Comment by njsmith** [comment]
Closing/opening to tickle CI, now that #905 is merged.

**Pull request merged**